### PR TITLE
reaverwps: enable parallel build

### PR DESCRIPTION
**Pull request by dasj19**
###### Motivation for this change
As a part of https://github.com/NixOS/nixpkgs/issues/142338 this change enables parallel builds for the package

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @trofi

### CODE REVIEW

**Comment by trofi** [comment]
On my 16-core machine the build fails consistently in parallel mode:

```
$ nix build -f. reaverwps
...
       last 10 log lines:
       > gcc -DCONF_DIR='"/var/db/reaver-wps-1.4/reaver"' -DREAVER_DATABASE='"/var/db/reaver-wps-1.4/reaver/reaver.db"' -Wall -g -O2 sigalrm.c -c
       > gcc -DCONF_DIR='"/var/db/reaver-wps-1.4/reaver"' -DREAVER_DATABASE='"/var/db/reaver-wps-1.4/reaver/reaver.db"' -Wall -g -O2 -Ilibwps sigint.c -c
       > gcc -DCONF_DIR='"/var/db/reaver-wps-1.4/reaver"' -DREAVER_DATABASE='"/var/db/reaver-wps-1.4/reaver/reaver.db"' -Wall -g -O2 iface.c -c
       > In file included from iface.h:40,
       >                  from iface.c:34:
       > lwe/iwlib.h:61:10: fatal error: wireless.h: No such file or directory
       >    61 | #include "wireless.h"
       >       |          ^~~~~~~~~~~~
       > compilation terminated.
       > make: *** [Makefile:67: iface.o] Error 1
```

Full build log: [reaver-wps-1.4-parallel.log.txt](https://github.com/NixOS/nixpkgs/files/7949655/reaver-wps-1.4-parallel.log.txt)

I think it will either require fixing header/binary dependencies upstream or explicitly disabling parallelism.

**Comment by dasj19** [comment]
okay, I am new to parallelism in nix, so maybe I am missing something,
both my machines succeed after adding `enableParallelBuilding = true;` to the package definition

**Comment by trofi** [comment]
> As for upstream ... the package is not maintained anymore

Aha, that might warrant for explicit `enableParallelBuilding = false;` with a comment (or a nixpkgs downstream fix, proposed below).

> okay, I am new to parallelism in nix, so maybe I am missing something, both my machines succeed after adding `enableParallelBuilding = true;` to the package definition

Yeah, that's the tricky part with parallel builds: they are frequently nondeterministic and execute build steps in slightly different order from run to run.

Sometime they work, sometimes they don't (depending on external machine load, available parallelism or even software bugs). Sometimes failures are very tricky to reproduce. I usually try to stress the build system with `--cores 64` or try to figure where to put `sleep` in the build system to widen race condition window. Also make sure your build log does enable parallelism (should contain something like `build flags: -j...`).

If we look at this specific example:

```
gcc -DCONF_DIR='"/var/db/reaver-wps-1.4/reaver"' -DREAVER_DATABASE='"/var/db/reaver-wps-1.4/reaver/reaver.db"' -Wall -g -O2 iface.c -c
In file included from iface.h:40,
                 from iface.c:34:
lwe/iwlib.h:61:10: fatal error: wireless.h: No such file or directory
   61 | #include "wireless.h"
      |          ^~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:67: iface.o] Error 1
```

we see that there seems to be a dependency between `iface.o` and `wireless.h`. If `wireless.h` is not present chances are it's autogenerated. Grepping for `wireless.h` we can find this:

```Makefile
# in lwe/Makefile
%: %.o
        $(CC) $(LDFLAGS) $(STRIPFLAGS) $(XCFLAGS) -o $@ $^ $(LIBS)
%.o: %.c wireless.h
        $(CC) $(XCFLAGS) -c $<
%.so: %.c wireless.h
        $(CC) $(XCFLAGS) $(PICFLAG) -c -o $@ $<
...
wireless.h:
        cp $(WEXT_HEADER) wireless.h
```

This looks OK. The only thing to note is `wireless.h` autogeneration. I bet if you add `sleep 10` in front of this `cp` you will get more reproducible build failure.

Let's look at `iface.o` build rule:

```Makefile
# in Makefile
...
iface.o: globule.o
        $(CC) $(CFLAGS) iface.c -c
```

Note how it's oblivious to the header dependency. You can reproduce build failure just by trying to build that target alone:

```
$ nix develop nixpkgs#reaverwps
$ dev>unpackPhase
$ dev>configurePhase
configure flags: --prefix=/tmp/zz/outputs/out --prefix=/tmp/zz/outputs/out --sysconfdir=/var/db/reaver-wps-1.4
...
$ dev> make iface.o
gcc -DCONF_DIR='"/var/db/reaver-wps-1.4/reaver"' -DREAVER_DATABASE='"/var/db/reaver-wps-1.4/reaver/reaver.db"' -Wall -g -O2 globule.c -c
gcc -DCONF_DIR='"/var/db/reaver-wps-1.4/reaver"' -DREAVER_DATABASE='"/var/db/reaver-wps-1.4/reaver/reaver.db"' -Wall -g -O2 iface.c -c
In file included from iface.h:40,
                 from iface.c:34:
lwe/iwlib.h:61:10: fatal error: wireless.h: No such file or directory
   61 | #include "wireless.h"
      |          ^~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:67: iface.o] Error 1
```

The failure looks close enough to original build failure.

Something like the below might be an OK fix for missing dependencies:

```diff
--- a/src/Makefile
+++ b/src/Makefile
@@ -63,7 +63,7 @@ session.o: globule.o
 80211.o: globule.o sql.o builder.o crc.o
        $(CC) $(CFLAGS) $(INC) 80211.c -c

-iface.o: globule.o
+iface.o: globule.o libiw
        $(CC) $(CFLAGS) iface.c -c

 sigalrm.o: globule.o
```

**Comment by dasj19** [comment]
Thanks for a comprehensive explanation, but now I wonder how could I fix something that does not brake on my machine... I see even ofborg bot compiles without failure

**Comment by trofi** [comment]
If you apply something like below locally I'd expect it to help in reproducing the failure:

```diff
--- a/lwe/Makefile
+++ b/lwe/Makefile
@@ ?,? ?,?
 wireless.h:
+    sleep 10
     cp $(WEXT_HEADER) wireless.h
 ```

**Comment by dasj19** [comment]
Ok, I was able to reproduce with the "sleep patch". I updated the PR.

Thanks a lot for guiding me into this, I'll try to apply the same heuristics on another package from the checklist

File: pkgs/tools/networking/reaver-wps/parallel-build.patch [hunk 1]
```diff
@@ -0,0 +1,11 @@
+--- a/src/Makefile.in	2012-01-18 10:02:40.000000000 +0100
```
Comment (trofi) [inline pkgs/tools/networking/reaver-wps/parallel-build.patch:1]: You'll probably need to apply the patch as well with something like `patches = [ ./parallel-build.patch ];`

**Review by trofi — CHANGES_REQUESTED** [review]

File: pkgs/tools/networking/reaver-wps/parallel-build.patch (hunk 1 above)
Comment (dasj19) [reply pkgs/tools/networking/reaver-wps/parallel-build.patch:1]: True, I lost that change when switching branches in git. Good you noticed. I'll change it right away.

Comment (dasj19) [reply pkgs/tools/networking/reaver-wps/parallel-build.patch:1]: Now should be okay.

**Review by trofi — APPROVED** [review]
Fixes parallel build for me on `x86_64-linux`.

**Pull request merged**